{"cells":[{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\n#for dirname, _, filenames in os.walk('/kaggle/input'):\n#    for filename in filenames:\n#        print(os.path.join(dirname, filename))\n\n# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":1,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from pathlib import Path\n\n# Path to train directory \ntrain_directory = Path('../input/chest-xray-pneumonia/chest_xray/chest_xray/train') \n\n# Path to validation directory\nval_directory = Path('../input/chest-xray-pneumonia/chest_xray/chest_xray/val')  \n\n# Path to test directory\ntest_directory = Path('../input/chest-xray-pneumonia/chest_xray/chest_xray/test')","execution_count":2,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from keras.preprocessing.image import ImageDataGenerator\n\n# create a data generator\ndatagen = ImageDataGenerator(rescale=1./255)","execution_count":3,"outputs":[{"output_type":"stream","text":"Using TensorFlow backend.\n","name":"stderr"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"# load and iterate training dataset\ntrain_it = datagen.flow_from_directory(train_directory, class_mode='binary', batch_size=64)\n# load and iterate validation dataset\nval_it = datagen.flow_from_directory(val_directory, class_mode='binary', batch_size=64)\n# load and iterate test dataset\ntest_it = datagen.flow_from_directory(test_directory, class_mode='binary', batch_size=64)","execution_count":4,"outputs":[{"output_type":"stream","text":"Found 5216 images belonging to 2 classes.\nFound 16 images belonging to 2 classes.\nFound 624 images belonging to 2 classes.\n","name":"stdout"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"from keras import layers\nfrom keras import models\nmodel = models.Sequential()\nmodel.add(layers.Conv2D(32, (3, 3), activation='relu',input_shape=(256, 256, 3)))\nmodel.add(layers.MaxPooling2D((2, 2)))\nmodel.add(layers.Conv2D(64, (3, 3), activation='relu'))\nmodel.add(layers.MaxPooling2D((2, 2)))\nmodel.add(layers.Conv2D(128, (3, 3), activation='relu'))\nmodel.add(layers.MaxPooling2D((2, 2)))\nmodel.add(layers.Conv2D(128, (3, 3), activation='relu'))\nmodel.add(layers.MaxPooling2D((2, 2)))\nmodel.add(layers.Flatten())\nmodel.add(layers.Dense(512, activation='relu'))\nmodel.add(layers.Dense(1, activation='sigmoid'))","execution_count":5,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#compile model using accuracy to measure model performance\nmodel.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])","execution_count":6,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#train the model\nmodel.fit_generator(train_it, steps_per_epoch=10, validation_data=val_it, validation_steps=50,epochs=20)","execution_count":7,"outputs":[{"output_type":"stream","text":"Epoch 1/10\n19/20 [===========================>..] - ETA: 1s - loss: 1.9532 - acc: 0.6799Traceback (most recent call last):\n  File \"/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3331, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-7-d49a0bd2feab>\", line 2, in <module>\n    model.fit_generator(train_it, steps_per_epoch=20, validation_data=val_it, validation_steps=50,epochs=10)\n  File \"/opt/conda/lib/python3.7/site-packages/keras/legacy/interfaces.py\", line 91, in wrapper\n    return func(*args, **kwargs)\n  File \"/opt/conda/lib/python3.7/site-packages/keras/engine/training.py\", line 1732, in fit_generator\n    initial_epoch=initial_epoch)\n  File \"/opt/conda/lib/python3.7/site-packages/keras/engine/training_generator.py\", line 242, in fit_generator\n    workers=0)\n  File \"/opt/conda/lib/python3.7/site-packages/keras/legacy/interfaces.py\", line 91, in wrapper\n    return func(*args, **kwargs)\n  File \"/opt/conda/lib/python3.7/site-packages/keras/engine/training.py\", line 1791, in evaluate_generator\n    verbose=verbose)\n  File \"/opt/conda/lib/python3.7/site-packages/keras/engine/training_generator.py\", line 365, in evaluate_generator\n    generator_output = next(output_generator)\n  File \"/opt/conda/lib/python3.7/site-packages/keras/utils/data_utils.py\", line 610, in get\n    inputs = future.get(timeout=30)\n  File \"/opt/conda/lib/python3.7/multiprocessing/pool.py\", line 651, in get\n    self.wait(timeout)\n  File \"/opt/conda/lib/python3.7/multiprocessing/pool.py\", line 648, in wait\n    self._event.wait(timeout)\n  File \"/opt/conda/lib/python3.7/threading.py\", line 552, in wait\n    signaled = self._cond.wait(timeout)\n  File \"/opt/conda/lib/python3.7/threading.py\", line 300, in wait\n    gotit = waiter.acquire(True, timeout)\nKeyboardInterrupt\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n    stb = value._render_traceback_()\nAttributeError: 'KeyboardInterrupt' object has no attribute '_render_traceback_'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/opt/conda/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1148, in get_records\n    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n  File \"/opt/conda/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 316, in wrapped\n    return f(*args, **kwargs)\n  File \"/opt/conda/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 350, in _fixed_getinnerframes\n    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n  File \"/opt/conda/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n  File \"/opt/conda/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n    filename = getsourcefile(frame) or getfile(frame)\n  File \"/opt/conda/lib/python3.7/inspect.py\", line 696, in getsourcefile\n    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n  File \"/opt/conda/lib/python3.7/inspect.py\", line 733, in getmodule\n    if ismodule(module) and hasattr(module, '__file__'):\n  File \"/opt/conda/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n    module = self._load()\n  File \"/opt/conda/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n    module = _importlib.import_module(self.__name__)\n  File \"/opt/conda/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 967, in _find_and_load_unlocked\n  File \"<frozen importlib._bootstrap>\", line 677, in _load_unlocked\n  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n  File \"/opt/conda/lib/python3.7/site-packages/tensorflow_core/python/keras/api/_v2/keras/__init__.py\", line 19, in <module>\n    from . import datasets\n  File \"/opt/conda/lib/python3.7/site-packages/tensorflow_core/python/keras/api/_v2/keras/datasets/__init__.py\", line 14, in <module>\n    from . import imdb\n  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 963, in _find_and_load_unlocked\n  File \"<frozen importlib._bootstrap>\", line 906, in _find_spec\n  File \"<frozen importlib._bootstrap_external>\", line 1280, in find_spec\n  File \"<frozen importlib._bootstrap_external>\", line 1252, in _get_spec\n  File \"<frozen importlib._bootstrap_external>\", line 1379, in find_spec\n  File \"<frozen importlib._bootstrap_external>\", line 59, in _path_join\nKeyboardInterrupt\n","name":"stdout"},{"output_type":"error","ename":"KeyboardInterrupt","evalue":"","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"]}]},{"metadata":{"trusted":true},"cell_type":"code","source":"test_datagen = ImageDataGenerator(rescale=1./255)\n\ntest_generator = test_datagen.flow_from_directory(\ntest_directory,\ntarget_size=(256, 256),\nbatch_size=64,\nclass_mode='binary')\ntest_loss, test_acc = model.evaluate_generator(test_generator, steps=10, verbose=1)\nprint('test acc:', test_acc)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}